File: 01_JAVA/49_EX/src/com/lion/view/SearchStudentInfoView.java
package com.lion.view;

import com.lion.controller.ControllerClass;
import com.lion.model.SoccerStudentClass;
import com.lion.model.StudentClass;
import com.lion.service.SearchStudentService;

import java.util.Scanner;

public class SearchStudentInfoView extends ViewClass{
    ControllerClass controllerClass;
    SearchStudentService searchStudentService;

    public SearchStudentInfoView(ControllerClass controllerClass){
        this.controllerClass = controllerClass;
        searchStudentService = new SearchStudentService();

/*        // 테스트용 데이터
        SoccerStudentClass soccerStudentClass1 = new SoccerStudentClass();
        soccerStudentClass1.inputStudentInfo();

        SoccerStudentClass soccerStudentClass2 = new SoccerStudentClass();
        soccerStudentClass2.inputStudentInfo();

        controllerClass.addStudentClass(soccerStudentClass1);
        controllerClass.addStudentClass(soccerStudentClass2);*/
    }

    @Override
    public void show() {
        System.out.println();

        // 리스트에 담겨져 있는 학생의 수를 가져온다.
        int stucentCount = controllerClass.getStudentCount();

        if(stucentCount == 0) {
            System.out.println("등록된 학생이 없습니다");
        } else {
            Scanner scanner = new Scanner(System.in);
            System.out.print("학생의 이름을 입력해주세요 : ");
            String searchStudentName = scanner.next();
            // 해당 이름의 학생이 있는지 검색한다.
            StudentClass studentClass = controllerClass.searchStudentByName(searchStudentName);
            // 학생이 없다면
            if(studentClass == null){
                System.out.println();
                System.out.println("검색된 학생이 없습니다");
            } else {
                studentClass.showStudentInfo();
            }
        }
    }
}

File: 02_Kotlin/53_StudentManager/src/com/lion/view/SearchStudentInfoView.kt
package com.lion.view

import com.lion.controller.ControllerClass
import com.lion.model.BaseBallStudentModel
import com.lion.model.BasketBallStudentModel
import com.lion.model.SoccerStudentModel
import com.lion.model.StudentModel
import com.lion.repository.StudentRepository
import com.lion.util.ProgramState
import com.lion.util.ToolClass
import java.util.Scanner

class SearchStudentInfoView(val controllerClass: ControllerClass) : ViewClass() {
    override fun show() {

        // 등록된 학생이 있는지 확인한다.
        val check1 = ToolClass.checkStudentExist()

        // 만약 학생이 없다면
        if(check1 == false){
            println()
            println("등록된 학생 정보가 없습니다.")
        }
        // 학생이 있다면
        else {
            // 학생 이름을 검색한다.
            val inputName = inputSearchStudentName()
            // 입력한 학생이 있는지 검사한다.
            val check2 = checkSearchStudentExist(inputName)
            // 입력한 학생이 없다면
            if(check2 == false){
                println()
                println("입력한 학생의 정보가 없습니다")
            }
            // 입력한 학생이 있다면
            else {
                // 검색한 학생의 정보를 가져온다.
                val searchStudentList = getSearchStudentInfo(inputName)
                // 가져온 학생의 수 만큼 반복한다.
                searchStudentList.forEach{
                    // 공통 부분 호출
                    it.showStudentInfo()
                    it.actionRun()
                    // 각각 다른 부분 호출
                    when(it){
                        // 축구라면
                        is SoccerStudentModel ->{
                            it.actionTakkle()
                        }
                        // 야구부라면
                        is BaseBallStudentModel -> {
                            it.actionStellBase()
                        }
                        // 농구부라면
                        is BasketBallStudentModel ->{
                            it.actionShutThreePoint()
                        }
                    }
                }
            }
        }

        // 프로그램 상태를 메인 메뉴로 변경한다.
        controllerClass.programState = ProgramState.STATE_SHOW_MAIN_MENU
    }

    // 학생 이름을 입력받는 기능
    private fun inputSearchStudentName():String{
        println()
        print("검색할 학생의 이름을 입력해주세요 : ")
        val scanner = Scanner(System.`in`)
        val inputName = scanner.next()
        return inputName
    }

    // 입력한 학생 정보가 있는지 확인하는 기능
    fun checkSearchStudentExist(inputName:String):Boolean{
        // 학생이 있는지 여부 값
        var isExistStudentName = false
        // 학생 데이터를 가져온다.
        val studentList = StudentRepository.getStudentInfo()
        // 학생의 수 만큼 반복한다.
        studentList.forEach {
            // 검색한 학생의 이름과 현재 학생 객체의 이름이 같은지 확인한다.
            if(inputName == it.studentName){
                // 변수에 true를 담아준다.
                isExistStudentName = true
                return@forEach
            }
        }
        return isExistStudentName
    }

    // 입력한 학생 정보를 가져오는 기능
    private fun getSearchStudentInfo(inputName:String):MutableList<StudentModel>{
        // 검색된 학생 정보를 담을 리스트
        val searchStudentList = mutableListOf<StudentModel>()
        // 학생 전체데이터를 가져온다.
        val studentList = StudentRepository.getStudentInfo()
        // 학생의 수 만큼 반복한다.
        studentList.forEach {
            // 입력한 학생의 이름과 현재 학생의 이름이 같다면..
            if(inputName == it.studentName){
                // 담아준다.
                searchStudentList.add(it)
            }
        }
        // 반환한다.
        return searchStudentList
    }
}







